### DISCUSSION TITLE
add to `Unsafe.SkipInit` , support for `ref struct` items

### DISCUSSION THREAD

**Discussion by irvnriir** [dotnet/csharplang#7581] [Language Ideas] [answered]
due to that `ref struct` can't be used as Generic 's Type Parameter, `Unsafe.SkipInit` doesn't work for it .

**Comment by huoyaoyuan** [comment]
Generic parameter support for ref struct is #1148.
`Unsafe.SkipInit` won't take effect for ref struct, since it can contain ref field, and it usually does contain.

**Reply by irvnriir** [reply]
this is not a suggestion for allowing general use as Generic parameter .

i do not see how this is related to ref fields .

**Reply by huoyaoyuan** [reply, marked as answer]
`Unsafe.SkipInit` is not known by the language. It is just a runtime-provided helper. Without generic parameter support, such helper can't be defined for ref structs.

Ref fields and reference type fields must be initialized to null since GC can't handle wild references.

**Reply by irvnriir** [reply]
i thought it has specific handling by compiler, to not error due to that it has not Definition for it's `out` parameter, and to avoid providing excessive output . https://github.com/dotnet/runtime/issues/29905 .

They don't exist until the struct is initialized, why/how would GC interact with them ?

**Reply by huoyaoyuan** [reply]
The definition order of a variable in source language is almost unrelated to the lifetime of its runtime representation. `out` variable definition is just a syntax sugar. You can see the [compilation results](https://sharplab.io/#v2:D4AQTAjAsAUCAMACEEB0AlArgOwC4EsBbAU1QGEB7QgB3wBtiAnAZSYDd8BjYgZwG5YsEAGZkYRGUQBvWIjnJRKJAFkIACgCU02fN0BVbDwCGAM1LMA1vmoBJbPlxqKmXInx5EADw0CYu3SAA7F6oACoUzLiM7gDmmr66AL46cikKyBAqYJrafv5y7q6eCfmIBsZmqJbWdg5OLl4+aQHBnmERUbHxackwiUA) to be undistinguishable.

**Reply by irvnriir** [reply]
If you mean that following not existing references is unsafe, thats the same for the struct itself .

**Reply by irvnriir** [reply]
> https://github.com/dotnet/csharplang/discussions/7581#discussioncomment-7215323

ik that, but don't see how its related

**Reply by irvnriir** [reply]
i meant compiler would error because it needs to ensure that its defined before the function ends .

**Reply by huoyaoyuan** [reply]
Yes, the same limitation also applies to structs with fields of reference types, like `ImmutableArray<T>`.

The real limitation is the usage pattern of the struct itself. `SkipInit` is designed for structs without managed fields. `ref struct` with no ref or managed fields should work in theory, but it doesn't need to be a ref struct then.

In conclusion, the method wouldn't provide meaningful benefit for a typical ref struct.

**Reply by irvnriir** [reply]
according to docs, `ref struct` is nothing more than a syntax to make compiler ensure that the struct doesn't go to Heap, logically that would mean that adding it to container in Heap would add a reference and not copy the struct . in that context, supporting `ref struct` without fields for managed reference types is what im asking for .

**Reply by huoyaoyuan** [reply]
Anything allowed in theory may be supported eventually, but things with more typical usage will be prioritized. For example, without generic parameter support for ref struct, `Action<Span<T>>` is not possible, but there's a specially declared `SpanAction<T>` due to its wide usage.

Support for any specific `ref struct` is not a language request and should go to dotnet/runtime. Support for all ref structs is blocked by generic parameter limitation. If we eventually relaxed the limitation, the request should go to dotnet/runtime to adopt the relax. Nothing can be done without working on generic parameter limitation now.

**Reply by theunrepentantgeek** [reply]
> according to docs, ref struct is nothing more than a syntax to make compiler ensure that the struct doesn't go to Heap

Not quite. Being located on the stack, it has a fixed lifetime: when the method exits, the storage is reclaimed.

Adding a reference to a heap based container would violate the lifetime rules - if the container lifetime exceeded the lifetime of the method, the reference would be stale.
